tests/test_pono.py

```diff
@@ -0,0 +1,175 @@
+import pytest
+import coreir
+import magma as m
+import smt_switch as ss
+from smt_switch.primops import (BVAdd, BVSub, BVMul, BVUdiv, Equal, Ite, Or,
+                                Distinct)
+import pono
+
+
+class ConfigReg(m.Circuit):
+    io = m.IO(D=m.In(m.Bits[2]), Q=m.Out(m.Bits[2])) + \
+        m.ClockIO(has_ce=True)
+
+    reg = m.Register(m.Bits[2], has_enable=True)(name="conf_reg")
+    io.Q @= reg(io.D, CE=io.CE)
+
+
+class SimpleALU(m.Circuit):
+    io = m.IO(a=m.In(m.UInt[16]),
+              b=m.In(m.UInt[16]),
+              c=m.Out(m.UInt[16]),
+              config_data=m.In(m.Bits[2]),
+              config_en=m.In(m.Enable),
+              ) + m.ClockIO()
+
+    opcode = ConfigReg(name="config_reg")(io.config_data, CE=io.config_en)
+    io.c @= m.mux([io.a + io.b, io.a - io.b, io.a * io.b, io.a / io.b], opcode)
+
+
+@pytest.mark.parametrize('config_data', range(4))
+def test_pono(config_data):
+    # Compile magma -> coreir
+    m.compile("build/SimpleALU", SimpleALU, output="coreir")
+
+    # Load compile result
+    context = coreir.Context()
+    context.load_library("commonlib")
+    top_mod = context.load_from_file("build/SimpleALU.json")
+
+    # Create solver/interpolator
+    solver = ss.create_msat_solver(False)
+    itp = ss.create_msat_interpolator()
+
+    # create RTS, populate with encoder
+    rts = pono.RelationalTransitionSystem(solver)
+    pono.CoreIREncoder(top_mod, rts)
+
+    # create BV sorts
+    bvsort16 = solver.make_sort(ss.sortkinds.BV, 16)
+    bvsort2 = solver.make_sort(ss.sortkinds.BV, 2)
+
+    # Reference to inputs
+    a = rts.named_terms["a"]
+    b = rts.named_terms["b"]
+
+    # Make state var equal to output (result of the implementation)
+    imp_res = rts.make_statevar('imp_res', bvsort16)
+    rts.assign_next(imp_res, rts.named_terms["self.c"])
+
+    # Create state var for spec result
+    spec_res = rts.make_statevar('spec_res', bvsort16)
+
+    # Expected op based on config_data param
+    op = [BVAdd, BVSub, BVMul, BVUdiv][config_data]
+
+    # Assign spec result to operation with a and b
+    rts.assign_next(
+        spec_res,
+        solver.make_term(op, a, b)
+    )
+
+    # State machine variable
+    test_state = rts.make_statevar('test_state', bvsort2)
+
+    # Starts at state 0
+    rts.constrain_init(solver.make_term(Equal, test_state,
+                                        solver.make_term(0, bvsort2)))
+
+    # sequential states 0 -> 1 -> 2 -> 3, then stays at 3
+    rts.assign_next(
```

Comment:
This could be:
```
rts.assign_next(test_state,
        rts.make_term(Ite,
                      solver.make_term(Equal, test_state, rts.make_term(3, bvsort2)),
                      rts.make_term(3, bvsort2),
                      rts.make_term(BVAdd, test_state, rts.make_term(1, bvsort2))))
```

But I don't think it makes much of a difference performance-wise -- just slightly less code. I will say that the way you wrote it is probably more realistic for the general case, i.e. if we get an arbitrary state machine, we might not be able to optimize it with a small encoding. Additionally, if it'd be helpful, I can add some kind of helper function to pono to generate this kind of encoding automatically. We'd just need to agree on an interface.

Reply:
Nice, yes that's more compact. I think for basic fault test cases they'll use this sequential/incrementing pattern, but once we add support for branches/loops it'll be a bit more complicated.  For an interface, perhaps some state machine object representation? (e.g. states, transitions) There are a few existing Python libraries for this, so perhaps we could leverage one of them as a representation and compile it down to Pono terms.  @phanrahan has used https://github.com/qntm/greenery with magma in the past and it seemed pretty nice (and simple)

Reply:
Cool, that makes a lot of sense! I'll look into greenery.